File: go/sketchkit/minhash/minhash.go
// Code authors: Vijay Erramilli and Codex
package minhash

import (
	"errors"
	"fmt"
	"math"

	sketchhash "github.com/llm-measurement/llm-sketchkit/go/sketchkit/hash"
	"github.com/llm-measurement/llm-sketchkit/go/sketchkit/internal/hashfamily"
	sketchpb "github.com/llm-measurement/llm-sketchkit/go/sketchkit/internal/pb"
	"google.golang.org/protobuf/proto"
)

const (
	wireVersion  = 1
	maxWireBytes = 4 << 20
)

// Profile names a MinHash profile from spec/profiles.md.
type Profile string

const (
	ProfileMicro   Profile = "micro"
	ProfileSmall   Profile = "small"
	ProfileDefault Profile = "default"
	ProfileK256    Profile = "k256"
)

var (
	// ErrUnknownProfile reports a profile outside spec/profiles.md.
	ErrUnknownProfile = errors.New("unknown minhash profile")

	// ErrInvalidSignatureLength reports a non-profile signature length.
	ErrInvalidSignatureLength = errors.New("invalid minhash signature length")

	// ErrIncompatibleMerge reports metadata mismatch on merge.
	ErrIncompatibleMerge = errors.New("incompatible minhash merge")

	// ErrCountOverflow reports populated-count overflow.
	ErrCountOverflow = errors.New("minhash populated count overflow")

	// ErrInvalidWireEncoding reports a malformed serialized sketch.
	ErrInvalidWireEncoding = errors.New("invalid minhash wire encoding")
)

var profileLengths = map[Profile]int{
	ProfileMicro:   64,
	ProfileSmall:   128,
	ProfileDefault: 128,
	ProfileK256:    256,
}

// Sketch is a mergeable MinHash signature over pre-hashed uint64 values.
type Sketch struct {
	profile   Profile
	length    int
	domain    sketchhash.Domain
	algorithm sketchhash.Algorithm

	signature      []uint64
	populatedCount uint64
}

// New constructs an empty MinHash sketch from a named profile.
func New(profile Profile, domain sketchhash.Domain, algorithm sketchhash.Algorithm) (*Sketch, error) {
	length, ok := profileLengths[profile]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownProfile, profile)
	}
	if !sketchhash.IsRegisteredDomain(domain) {
		return nil, fmt.Errorf("%w: %s", sketchhash.ErrUnregisteredDomain, domain)
	}
	if algorithm != sketchhash.HMACSHA25664 {
		return nil, fmt.Errorf("%w: %s", ErrIncompatibleMerge, algorithm)
	}

	return newSketch(profile, length, domain, algorithm), nil
}

func newSketch(
	profile Profile,
	length int,
	domain sketchhash.Domain,
	algorithm sketchhash.Algorithm,
) *Sketch {
	signature := make([]uint64, length)
	for i := range signature {
		signature[i] = math.MaxUint64
	}

	return &Sketch{
		profile:   profile,
		length:    length,
		domain:    domain,
		algorithm: algorithm,
		signature: signature,
	}
}

// AddHash updates the signature with a pre-hashed value.
func (s *Sketch) AddHash(hash uint64) error {
	if s.populatedCount == math.MaxUint64 {
		return ErrCountOverflow
	}
	for i := range s.signature {
		value := hashfamily.MinHash(hash, uint32(i))
		if value < s.signature[i] {
			s.signature[i] = value
		}
	}
	s.populatedCount++

	return nil
}

// Merge applies element-wise minimum with other. Metadata must match.
func (s *Sketch) Merge(other *Sketch) error {
	if other == nil {
		return nil
	}
	if s.profile != other.profile ||
		s.length != other.length ||
		s.domain != other.domain ||
		s.algorithm != other.algorithm {
		return fmt.Errorf("%w: metadata differs", ErrIncompatibleMerge)
	}
	if s.populatedCount > math.MaxUint64-other.populatedCount {
		return ErrCountOverflow
	}
	for i, value := range other.signature {
		if value < s.signature[i] {
			s.signature[i] = value
		}
	}
	s.populatedCount += other.populatedCount

	return nil
}

// JaccardEstimate returns the MinHash Jaccard estimate against other.
func (s *Sketch) JaccardEstimate(other *Sketch) (float64, error) {
	if other == nil {
		return 0, fmt.Errorf("%w: nil", ErrIncompatibleMerge)
	}
	if s.profile != other.profile ||
		s.length != other.length ||
		s.domain != other.domain ||
		s.algorithm != other.algorithm {
		return 0, fmt.Errorf("%w: metadata differs", ErrIncompatibleMerge)
	}
	if s.populatedCount == 0 && other.populatedCount == 0 {
		return 1, nil
	}
	if s.populatedCount == 0 || other.populatedCount == 0 {
		return 0, nil
	}

	matches := 0
	for i, value := range s.signature {
		if value == other.signature[i] {
			matches++
		}
	}

	return float64(matches) / float64(s.length), nil
}

// PopulatedCount returns the number of AddHash calls observed by this sketch.
func (s *Sketch) PopulatedCount() uint64 {
	return s.populatedCount
}

// SignatureLength returns the configured signature length.
func (s *Sketch) SignatureLength() int {
	return s.length
}

// Signature returns a copy of the current signature.
func (s *Sketch) Signature() []uint64 {
	out := make([]uint64, len(s.signature))
	copy(out, s.signature)

	return out
}

// RepresentationMode returns the current protobuf representation mode.
func (s *Sketch) RepresentationMode() sketchpb.RepresentationMode {
	return sketchpb.RepresentationMode_REPRESENTATION_MODE_MINHASH_SIGNATURE
}

// Clone returns a deep copy of s.
func (s *Sketch) Clone() *Sketch {
	clone := newSketch(s.profile, s.length, s.domain, s.algorithm)
	clone.populatedCount = s.populatedCount
	copy(clone.signature, s.signature)

	return clone
}

// MarshalBinary returns deterministic protobuf bytes.
func (s *Sketch) MarshalBinary() ([]byte, error) {
	message := s.toProto()
	return proto.MarshalOptions{Deterministic: true}.Marshal(message)
}

// Parse decodes a MinHash sketch from deterministic protobuf bytes.
func Parse(data []byte) (*Sketch, error) {
	if len(data) > maxWireBytes {
		return nil, fmt.Errorf("%w: input length %d exceeds %d", ErrInvalidWireEncoding, len(data), maxWireBytes)
	}

	var message sketchpb.Sketch
	if err := proto.Unmarshal(data, &message); err != nil {
		return nil, err
	}

	return fromProto(&message)
}

func (s *Sketch) toProto() *sketchpb.Sketch {
	length := uint32(s.length)
	return &sketchpb.Sketch{
		Metadata: &sketchpb.SketchMetadata{
			Kind:                   sketchpb.SketchKind_SKETCH_KIND_MINHASH,
			WireVersion:            wireVersion,
			Profile:                string(s.profile),
			HashDomain:             string(s.domain),
			HashAlgo:               sketchpb.HashAlgorithm_HASH_ALGORITHM_HMAC_SHA256_64,
			RepresentationMode:     s.RepresentationMode(),
			MinhashSignatureLength: &length,
		},
		Body: &sketchpb.Sketch_Minhash{
			Minhash: &sketchpb.MinHashSketch{
				Signature:      append([]uint64(nil), s.signature...),
				PopulatedCount: s.populatedCount,
			},
		},
	}
}

func fromProto(message *sketchpb.Sketch) (*Sketch, error) {
	metadata := message.GetMetadata()
	body := message.GetMinhash()
	if metadata == nil || body == nil {
		return nil, fmt.Errorf("%w: missing minhash metadata or body", ErrInvalidWireEncoding)
	}
	if metadata.GetKind() != sketchpb.SketchKind_SKETCH_KIND_MINHASH {
		return nil, fmt.Errorf("%w: kind %s", ErrInvalidWireEncoding, metadata.GetKind())
	}
	if metadata.GetWireVersion() != wireVersion {
		return nil, fmt.Errorf("%w: wire version %d", ErrInvalidWireEncoding, metadata.GetWireVersion())
	}
	if metadata.GetHashAlgo() != sketchpb.HashAlgorithm_HASH_ALGORITHM_HMAC_SHA256_64 {
		return nil, fmt.Errorf("%w: hash algorithm %s", ErrInvalidWireEncoding, metadata.GetHashAlgo())
	}
	if metadata.GetRepresentationMode() != sketchpb.RepresentationMode_REPRESENTATION_MODE_MINHASH_SIGNATURE {
		return nil, fmt.Errorf("%w: representation %s", ErrInvalidWireEncoding, metadata.GetRepresentationMode())
	}

	profile := Profile(metadata.GetProfile())
	length, ok := profileLengths[profile]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownProfile, profile)
	}
	if metadata.GetMinhashSignatureLength() != uint32(length) {
		return nil, fmt.Errorf("%w: profile %s length=%d",
			ErrInvalidSignatureLength, profile, metadata.GetMinhashSignatureLength())
	}
	if !sketchhash.IsRegisteredDomain(sketchhash.Domain(metadata.GetHashDomain())) {
		return nil, fmt.Errorf("%w: %s", sketchhash.ErrUnregisteredDomain, metadata.GetHashDomain())
	}
	if len(body.GetSignature()) != length {
		return nil, fmt.Errorf("%w: signature length", ErrInvalidWireEncoding)
	}
	if body.GetPopulatedCount() == 0 {
		for _, value := range body.GetSignature() {
			if value != math.MaxUint64 {
				return nil, fmt.Errorf("%w: non-empty signature with zero count", ErrInvalidWireEncoding)
			}
		}
	}

	sketch := newSketch(profile, length, sketchhash.Domain(metadata.GetHashDomain()), sketchhash.HMACSHA25664)
	sketch.populatedCount = body.GetPopulatedCount()
	copy(sketch.signature, body.GetSignature())

	return sketch, nil
}

File: python/llm_sketchkit/minhash.py
# SPDX-License-Identifier: Apache-2.0
# Code authors: Vijay Erramilli and Codex
"""Pure-Python MinHash sketch implementation."""

from __future__ import annotations

from typing import Any, cast

from . import _proto, hashfamily, profiles

MAX_UINT64 = (1 << 64) - 1


class MinHashError(ValueError):
    """Base class for MinHash failures."""


class UnknownProfileError(MinHashError):
    """Raised when a profile is not supported."""


class InvalidSignatureLengthError(MinHashError):
    """Raised when wire metadata uses a non-profile signature length."""


class IncompatibleMergeError(MinHashError):
    """Raised when merge metadata differs."""


class CountOverflowError(MinHashError):
    """Raised when populated count exceeds uint64."""


class InvalidWireEncodingError(MinHashError):
    """Raised when serialized bytes are malformed."""


class Sketch:
    """Mergeable MinHash signature over pre-hashed unsigned 64-bit values."""

    def __init__(
        self,
        profile: str,
        domain: str = profiles.PROMPT_V1,
        algorithm: str = profiles.HMAC_SHA256_64,
    ) -> None:
        length = profiles.MINHASH_SIGNATURE_LENGTHS.get(profile)
        if length is None:
            raise UnknownProfileError(profile)
        if domain not in profiles.REGISTERED_DOMAINS:
            raise IncompatibleMergeError(domain)
        if algorithm != profiles.HMAC_SHA256_64:
            raise IncompatibleMergeError(algorithm)

        self._profile = profile
        self._length = length
        self._domain = domain
        self._algorithm = algorithm
        self._signature = [MAX_UINT64] * length
        self._populated_count = 0

    @classmethod
    def parse(cls, data: bytes) -> Sketch:
        """Decode a MinHash sketch from deterministic protobuf bytes."""

        if len(data) > _proto.MAX_WIRE_BYTES:
            raise InvalidWireEncodingError("wire input too large")
        message = _proto.parse_sketch(data)
        metadata = message.metadata
        if not message.HasField("metadata") or not message.HasField("minhash"):
            raise InvalidWireEncodingError("missing MinHash metadata or body")
        if metadata.kind != _proto.SKETCH_KIND_MINHASH:
            raise InvalidWireEncodingError("wrong sketch kind")
        if metadata.wire_version != _proto.WIRE_VERSION:
            raise InvalidWireEncodingError("wrong wire version")
        if metadata.hash_algo != _proto.HASH_ALGORITHM_HMAC_SHA256_64:
            raise InvalidWireEncodingError("wrong hash algorithm")
        if metadata.representation_mode != _proto.REPRESENTATION_MINHASH_SIGNATURE:
            raise InvalidWireEncodingError("wrong representation")

        profile = cast(str, metadata.profile)
        length = profiles.MINHASH_SIGNATURE_LENGTHS.get(profile)
        if length is None:
            raise UnknownProfileError(profile)
        if cast(int, metadata.minhash_signature_length) != length:
            raise InvalidSignatureLengthError(profile)
        if cast(str, metadata.hash_domain) not in profiles.REGISTERED_DOMAINS:
            raise IncompatibleMergeError("unregistered domain")

        body = message.minhash
        signature = [cast(int, value) for value in body.signature]
        populated_count = cast(int, body.populated_count)
        if len(signature) != length:
            raise InvalidWireEncodingError("wrong signature length")
        if populated_count == 0 and any(value != MAX_UINT64 for value in signature):
            raise InvalidWireEncodingError("non-empty signature with zero count")

        sketch = cls(profile, cast(str, metadata.hash_domain), profiles.HMAC_SHA256_64)
        sketch._signature = signature
        sketch._populated_count = populated_count
        return sketch

    def add_hash(self, value: int) -> None:
        """Update the signature with a pre-hashed unsigned 64-bit value.

        CountOverflowError leaves self unchanged. Runtime failures and inputs
        outside the documented types have no rollback guarantee.
        """

        if self._populated_count == MAX_UINT64:
            raise CountOverflowError("populated count")
        value &= hashfamily.MASK64
        for index in range(self._length):
            candidate = hashfamily.minhash(value, index)
            if candidate < self._signature[index]:
                self._signature[index] = candidate
        self._populated_count += 1

    def merge(self, other: Sketch | None) -> None:
        """Merge another MinHash sketch with identical metadata.

        IncompatibleMergeError and CountOverflowError leave self unchanged.
        None is a no-op; a distinct source is never modified. Runtime failures,
        invalid private state, and concurrent use have no rollback guarantee.
        """

        if other is None:
            return
        if (
            self._profile != other._profile
            or self._length != other._length
            or self._domain != other._domain
            or self._algorithm != other._algorithm
        ):
            raise IncompatibleMergeError("metadata differs")
        if self._populated_count > MAX_UINT64 - other._populated_count:
            raise CountOverflowError("populated count")
        self._signature = [
            min(left, right)
            for left, right in zip(self._signature, other._signature, strict=True)
        ]
        self._populated_count += other._populated_count

    def jaccard_estimate(self, other: Sketch) -> float:
        """Return the MinHash Jaccard estimate against another sketch."""

        if (
            self._profile != other._profile
            or self._length != other._length
            or self._domain != other._domain
            or self._algorithm != other._algorithm
        ):
            raise IncompatibleMergeError("metadata differs")
        if self._populated_count == 0 and other._populated_count == 0:
            return 1.0
        if self._populated_count == 0 or other._populated_count == 0:
            return 0.0
        matches = sum(
            1
            for left, right in zip(self._signature, other._signature, strict=True)
            if left == right
        )
        return float(matches) / float(self._length)

    def populated_count(self) -> int:
        """Return the number of AddHash calls observed."""

        return self._populated_count

    def signature_length(self) -> int:
        """Return the configured signature length."""

        return self._length

    def signature(self) -> list[int]:
        """Return a copy of the signature words."""

        return list(self._signature)

    def signature_hex(self) -> list[str]:
        """Return signature words as fixed-width lowercase hex strings."""

        return [f"{value:016x}" for value in self._signature]

    def clone(self) -> Sketch:
        """Return a deep copy of the sketch."""

        clone = type(self)(self._profile, self._domain, self._algorithm)
        clone._signature = list(self._signature)
        clone._populated_count = self._populated_count
        return clone

    def marshal_binary(self) -> bytes:
        """Return deterministic protobuf bytes."""

        return _proto.serialize(self._to_proto())

    def _to_proto(self) -> Any:
        metadata = _proto.sketchpb.SketchMetadata(
            kind=_proto.SKETCH_KIND_MINHASH,
            wire_version=_proto.WIRE_VERSION,
            profile=self._profile,
            hash_domain=self._domain,
            hash_algo=_proto.HASH_ALGORITHM_HMAC_SHA256_64,
            representation_mode=_proto.REPRESENTATION_MINHASH_SIGNATURE,
        )
        metadata.minhash_signature_length = self._length

        body = _proto.sketchpb.MinHashSketch(
            signature=self._signature,
            populated_count=self._populated_count,
        )
        return _proto.sketchpb.Sketch(metadata=metadata, minhash=body)


def new(
    profile: str,
    domain: str = profiles.PROMPT_V1,
    algorithm: str = profiles.HMAC_SHA256_64,
) -> Sketch:
    """Construct an empty MinHash sketch."""

    return Sketch(profile, domain, algorithm)


def parse(data: bytes) -> Sketch:
    """Decode a MinHash sketch from deterministic protobuf bytes."""

    return Sketch.parse(data)
